### DISCUSSION TITLE
Ability to generate PDF invioces that can be emailed to customers or downloaded from admin

### DISCUSSION THREAD

**Discussion by mrguruo** [medusajs/medusa#4719] [Feature Requests]
Many customers want a PDF invoice, especially for B2B sales, for their accounting records. Also many stores include a hardcopy invoice with the shipment. Having a PDF generated invoice and packing slip would be helpful. Similar Shopify apps that do this:
https://apps.shopify.com/avada-pdf-invoice
https://apps.shopify.com/invoice-wizard

**Comment by hexrw** [comment]
This feature is crucial for my use case

**Reply by amns23** [reply]
Same for me, it's a required feature for e-commerce in France.
I hope it will be available soon, in the meantime there may be independent tools available ✌️

**Reply by cmario92** [reply]
yes that would be nice, it is a must have feature!

**Reply by radoslaw-sz** [reply]
hi @cmario92  - you may want to check: https://github.com/RSC-Labs/medusa-documents

**Comment by radoslaw-sz** [comment]
Hi all - probably my company will take it soon. Stay tuned :)

**Comment by radoslaw-sz** [comment]
hi @mrguruo , @hexrw  and @amns23 
I am looking for some information related to invoice and packings slips (or maybe other functionalities).
Could you please consider some very basic scenario, how you would like to work? Right now I am thinking about list of orders for which you can click on "generate" and get the invoice and packing slip. In the next step I would see clicking on the "send" button and send invoice to customer.
What else basic functionalities you see? I am not an expert in e-commerce, so I am wondering what is the most common scenarios which we need to cover here.

**Reply by hexrw** [reply]
In the existing order route, add a new button using custom admin widgets which will use the existing browser print API to print out a dynamically generated PDF file containing only information related to delivery, this "packaging slip" is basically what you'd put on a package 📦 you're sending to your friend through the post office 🏤. Then add other buttons like sending invoices to the customer (ideally supporting the Stripe integration as that's what I believe most people use for payments anyway). You can also add anything else you think would be useful like inventory cards etc. which would be inside products and/or inventory and again would he a document for printing 🖨️.

Let the user create is own HTML templates for these PDF documents by putting them inside `data/`. Use handlebars and a library like [jsPDF](https://github.com/parallax/jsPDF) to then generate these documents. Take inspiration from how the templating works in [medusa-plugin-ses](https://medusajs.com/plugins/medusa-plugin-ses/).

I was planning on adding this feature myself but unfortunately I don't have the time so if you're willing to take this on I'm happy to help.

**Reply by radoslaw-sz** [reply]
thanks for a great response!
However, I made some small different assumption as we cannot easily modify orders list. I made separated sidebar button with "Invoices" which will contain all actions related to that, including modified version of orders list.
Right now I am focusing on one basic template working with https://react-pdf.org/.  Unfortunately, customization of template is far away from what I plan to do, I won't be able to implement it soon...
One thing is interesting what you mentioned - Stripe integration. From what I know, invoice can be sent to customer even before payment - am I wrong? I was thinking that generation of invoice can be done for any order, paid or not paid yet. Also same for package slip - sometimes we may ship something and then get payment. I am struggling whether we need here some connection between statuses of order and invoice functionality itself.

**Reply by radoslaw-sz** [reply]
In short, I am thinking about something like this (more or less) 
![image](https://github.com/medusajs/medusa/assets/54472480/521e3da2-112a-4c0c-ad14-e485ef30a6d9)

**Reply by hexrw** [reply]
That seems great. Would be very useful. You could support Stripe, that would be even much easier then manual payment provider, as Stripe gives you all invoices through a simple REST API.

What I personally need the most is the packaging slip functionality though. If you could pull that off that would be awesome and I'd be very happy to help.

Please reach out to me, there's more plugins I wanted to work on and it would be great if we could collaborate. Here's my email (aliased): [hexrw.arming310@passinbox.com](mailto:hexrw.arming310@passinbox.com)

**Comment by radoslaw-sz** [comment]
hi @mrguruo , @hexrw , @amns23 - the very first version of plugin is available here: https://github.com/RSC-Labs/medusa-documents (or on NPM: https://www.npmjs.com/package/@rsc-labs/medusa-documents) . I will appreciate any feedback (I encourage to reach me on Discord, where I will give the updates).

**Reply by hexrw** [reply]
Could you send me your Discord username on the email I provided so that I could contact you?

**Reply by radoslaw-sz** [reply]
@hexrw sent
